README.md
# Mod Config Reborn

**Gagarin Patch Compatible**

Mod Config Reborn is the next generation version of WayWockets original Mod Config.
Built as a team effort to keep this essential mod alive on the workshop through this and future patch releases.

Mod Config Reborn cannot be used with Surviving Mars versions less than Gagarin (Space Race)

#### For the Player
Install this mod to get additional configuration options for mods that support it.

#### For the Modder
Use this mod to create user settable options for your mods.
Gagarin (Space Race) is significantly different from previous versions of the game.
Read the user/modder guide to see what needs to be changed to get this version working with your your mod.
You will need to make minor adjustments otherwise Mod Config will not work properly with your mod.

Guide Link > [Modders Guide](https://steamcommunity.com/sharedfiles/filedetails/?id=1543466595)

#### Legacy Data
This version will not carry over any of the original Mod Config data settings.
All settings will default to the mods defaults upon running this mod.

#### Acknowledgements
A great big shout out and hearty acknowledgement goes to Waywocket, who built Mod Config way back when Surviving Mars first came out.
His foresight into the modding realm allowed modders such as myself and many others to offer "adjustment parameters" to their own mods making them rich with features that otherwise would be static in nature.
Since the inception, this mod has been an essential part of Surviving Mars and the modder community, and on behalf of the modder community, thank you for your work.

Thank you ChoGGi for your great coding skills leant to this project

This is a team project composed of Waywocket, SkiRich and ChoGGi.

Use the Need Help discussion in this mod for help or find us on discord.
[Surviving Mars Discord](https://discord.gg/eS2Nap)

Report bugs in the Bug Report discussion in this mod.

Dont forget to VOTE UP!

### Other Links
- [Mod Config Reborn on Github](https://github.com/SkiRich/Mod_Config_Reborn)
- [Mod Config Reborn on Steam](https://steamcommunity.com/sharedfiles/filedetails/?id=1542863522)
- Nexus - When Gagarin Reborn is no longer a release candidate - before Gararin Patch rolls.

### Some interesting Collections:
- [SkiRich's Mod Collection](https://steamcommunity.com/workshop/filedetails/?id=1418440299)
- [ChoGGi's Mod Collection](https://steamcommunity.com/workshop/filedetails/?id=1411210466)





CHANGELOG.md
# Mod Config Reborn Changelog
## v2.6.0 09/05/2021 9:41:32 PM
#### Changed
- Versioning bump for Picard

#### Added

#### Removed

#### Fixed Issues

#### Open Issues

#### Deprecated

#### Todo

--------------------------------------------------------
## v2.5.5 04/04/2021 2:27:18 PM
#### Changed
- MCR_docs.md file WaitForModConfig

--------------------------------------------------------
## 2.5.4 03/16/2021 8:11:58 PM
#### Changed
- Text in main menu to not confuse with in game mod manager

--------------------------------------------------------
## 2.5.3 06/16/2019 2:19:10 AM
#### Changed
- All font styles updated to use current text styles
- changed StringIDBase to my range
- moved starting global vars to MCR_1Classes file

#### Fixed Issues
- font used did not paint russian cirillic characters

--------------------------------------------------------
## v2.5.2 06/05/2019 7:58:40 PM
#### Changed
- ModConfig:Load()
  - now added err to ModLog
  - Does an immediate write to blank out the persistent data in account.dat
- ModConfig.CloseDialog(dlg)
  - using Close() instead of delete()
- changed various ModConfig.something to self.something
- minor changes to retrofit routine and using table.find

#### Removed
- OnSetRollover = function from local close_button
- close_button.OnPress = ModConfig.CloseDialog

#### Fixed Issues
- Interpolation issues generated when closing the modconfig dialog

--------------------------------------------------------
## v2.5.1 06/04/2019 9:29:07 PM
#### Changed
- ModConfig:CalcDataSpace()

#### Fixed Issues
- CalcDataSpace() was calculating incorrect srting length size

--------------------------------------------------------
## v2.5.0 06/04/2019 1:01:53 AM
#### Changed
- function ModConfig:ReadSettingsFile(restore)
  - changed the way it reads the file data, allows for two types of reads, restore and localstorage
  - no longer walking registry
  - removed convertion techniques, not needed here
- function ModConfig:CalcDataSpace(compressData)
  - no longer calculating compression by default
  - no longer walking registry
- function ModConfig:SaveSettingsFile(backup)
  - changed the way it saves the file data, allows for two types of save, backup and localstorage
  - no longer walking registry
  - removed convertion techniques, not needed here
- ModConfig:Save()
  - no longer walking registry
  - no longer compressing
- ModConfig:Load()
  - no longer walking registry
  - no longer decompressing

#### Added
- Additional ModLog messages

#### Removed
- Removed initialize LocalStorage, moved to read and save file functions.
- no longer compressing data
- removed OnMsg.CityStart() , LoadGame(), and CityStart() - didnt need them.

#### Fixed Issues
- registry walking in the load and save functions was flawed after the game converted to
using double load of mods (ReloadLua()), once during game load on the start screen and again on game load/new.
The result was an incomplete registry since not all mods loaded yet but some mods would do a ModConfig:Set().
Walking the registry early caused some mod data to be wiped clean on every game load or new game.

#### Todo
- Add a compress technique and modconfig optoin to use if data size creep gets too large.

--------------------------------------------------------
## v2.4.5 06/01/2019 3:58:11 AM

#### Added
- OnMsg.CityStart

#### Fixed Issues
- ReadPersistentStorage was wiping the storage on citystart and exit.

--------------------------------------------------------
## v2.4.4 05/25/2019 4:40:29 AM

#### Added
- added OnMsg.SaveGame()
- added OnMsg.LoadGame()

#### Removed

#### Fixed Issues
- ReadModPersistentData() was clearing persistent data upon being issued causing write data not to be current, so if a user loaded a game and quit the data was reset on next load.

--------------------------------------------------------
## v2.4.3 05/09/2019 10:44:35 PM

#### Added
- Added g_ModConfigLoadedInitData
  - a variable that contains a copy of the registry at the moment MCR loads and before any other mod makes changes
  - used for debugging
- Added additional prints and ModLog's to code when MCR fails during load - for debugging

--------------------------------------------------------
## v2.4.2 12/11/2018 5:05:37 PM
#### Changed
- ModConfig:Save()
- ModConfig:RegisterMod()
- ModConfig:ResetAllToDefaults() now emulates the Set() command but forces the set
- ModConfig:Save() - added reset parm for passthrought to SaveSettingsFile()
- ModConfig:SaveSettingsFile(0 - added reset param for special handling of defaults

#### Added
- parameter restrictions for sav_loc in RegisterMod()
- token ModConfigReset send in Msg ModConfigChanged in ResetAllToDefaults()
- check for persistent storage limit reached in ModConfig:CalcDataSpace()
- ModConfigWarnOverLimit() - a check for persistent storage space prior to save.

#### Fixed Issues
- fixed check for nil in Save()

--------------------------------------------------------
## v2.4.1 12/03/2018 11:23:09 PM
#### Changed
- fixed syntax and documentation of comments

--------------------------------------------------------
## v2.4.0 12/03/2018 8:08:45 PM
#### Changed
- function ModConfig:RegisterMod
  - included option for sav_loc
- ModConfig:CalcDataSpace(mod_data)
  - sending save_data from Save() function to caculate only the central storage
- ModConfig:Save()
  - now splits save to central or local storage
- ModConfig:Load()
  - now calls ReadSettingsFile() and consolidates the data in ModConfig.data
- ModConfig:CalcDataSpace()
  - takes into account local vs central storage and returns only space used by central storage.

#### Added
- Added functions to read and save to LocalStorage.lua
- function ModConfig:ReadSettingsFile()
- function ModConfig:SaveSettingsFile()


Code/MCR_docs.lua

--[[
    Handling messages sent by this mod:

    function OnMsg.UIReady()
        Sent once the game's UI has loaded and is ready to query or modify.

    function OnMsg.ModConfigReady()
        Sent once ModConfig has finished loading and it's safe to start using.

    function OnMsg.ModConfigChanged(mod_id, option_id, value, old_value, token)
        Sent whenever any mod option is changed.
        The 'token' parameter matches the token given to ModConfig:Set(). The intention here is to
        make it easier for you to filter messages you shouldnt be responding to; if you set an
        option yourself you might want to pass in a token so that your handler can check for it and
        ignore the message.
--]]


--[[  Use this code to detect modconfig on startup

local mod_name = "@MOD" -- Replace @MOD with your mods name.  It will show up in the logs.
local ModConfig_id = "1340775972"
local ModConfigWaitThread = false
g_ModConfigLoaded = table.find_value(ModsLoaded, "steam_id", ModConfig_id) or false


-- wait for mod config to load or fail out and use defaults
local function WaitForModConfig()
	if (not ModConfigWaitThread) or (not IsValidThread(ModConfigWaitThread)) then
		ModConfigWaitThread = CreateRealTimeThread(function()
	    if lf_print then print(string.format("%s WaitForModConfig Thread Started", mod_name)) end
      local threadlimit = 120  -- loops to wait before fail and exit thread loop
      local runloop     = true -- var to exit early
 		  while runloop and threadlimit > 0 do
 			  g_ModConfigLoaded = table.find_value(ModsLoaded, "steam_id", ModConfig_id) or false
 			  if g_ModConfigLoaded and ModConfig:IsReady() then
 			  	runloop = false
 			  else
 			    Sleep(500) -- Sleep 1/2 second
 			  end -- if g_ModConfigLoaded
 			  threadlimit = threadlimit - 1
 		  end -- while
      if lf_print then print(string.format("%s WaitForModConfig Thread Continuing", mod_name)) end

      -- See if ModConfig is installed and any defaults changed
      if g_ModConfigLoaded and ModConfig:IsReady() then

      	-- PUT CODE HERE --

      else
      	-- PUT DEFAULTS HERE OR SET THEM UP BEFORE RUNNING THIS FUNCTION ---

    	  if lf_print then print(string.format("**** %s - Mod Config Never Detected On Load - Using Defaults ****", mod_name)) end
    	  ModLog(string.format("**** %s - Mod Config Never Detected On Load - Using Defaults ****", mod_name))
      end -- end if g_ModConfigLoaded
      if lf_print then print(string.format("%s WaitForModConfig Thread Ended", mod_name)) end
		end) -- thread
	else
		if lf_print then print(string.format("%s Error - WaitForModConfig Thread Never Ran", mod_name)) end
		ModLog(string.format("%s Error - WaitForModConfig Thread Never Ran", mod_name))
	end -- check to make sure thread not running
end -- WaitForModConFig

--]]


Code/MCR_1Classes.lua
-- Code developed for Mod Config Reborn
-- Code originally developed by Waywocket as Mod Config
-- Modified to work with Gagarin Patch
-- Re-written to optimize code
-- Author @SkiRich
-- This mod is subject to the terms and conditions outlined in the LICENSE file included in this mod.
-- Created Oct 14th, 2018
-- Updated March 16th, 2021

local lf_debug   = false  -- used only for certain ex() instance
local lf_print   = false  -- Setup debug printing in local file
                          -- Use if lf_print then print("something") end

ModConfig = {}    -- base class for modconfig
ModConfig.StringIdBase = 17764703500 -- Mod Config Reborn    : 703500 - 703599; This file 0-20
g_ModConfigLoaded = false -- set detection mechanism var. Set true in ClassesGenerate()
g_ModConfigLoadedInitData = false -- test variable for initially loaded data
g_ModConfigLoadedRegistry = false -- test variable for initially loaded registry


----------------------------------------------------------------------------------------------------------

DefineClass.XModConfigToggleButton = {
    __parents = {
        "XToggleButton"
    },
    properties = {
        { category = "Image", id = "DisabledImage", editor = "text", default = "UI/Icons/traits_approve_disable.tga" },
        { category = "Image", id = "EnabledImage",  editor = "text", default = "UI/Icons/traits_approve.tga" },
        { category = "Params", id = "ModId",        editor = "text", default = "" },
        { category = "Params", id = "OptionId",     editor = "text", default = "" },
    },
    Dock = "right",
    HAlign = "right",
    VAlign = "center",
    TextFont = "OnScreenHintText",
    TextStyle = "OnScreenHintText",
    Background = RGBA(0, 0, 0, 0),
    RolloverBackground = RGBA(0, 0, 0, 0),
    PressedBackground = RGBA(0, 0, 0, 0),
    MouseCursor = "UI/Cursors/Rollover.tga",
}
function XModConfigToggleButton:Init()
    self.idIcon.ImageFit = "height"
    self:SetToggled(ModConfig:Get(self.ModId, self.OptionId), false)
end

function XModConfigToggleButton:SetToggled(toggled, update)
    self:Set(toggled, update)
end

function XModConfigToggleButton:Set(value, update)
    if update == nil then update = true end
    self.Toggled = value
    self.idIcon:SetImage(value and self.EnabledImage or self.DisabledImage)
    if update then
        ModConfig:Set(self.ModId, self.OptionId, value, ModConfig.internal_token)
    end
end

----------------------------------------------------------------------------------------------------------
DefineClass.XModConfigEnum = {
    __parents = {
        "XPageScroll",
    },
    properties = {
        { category = "Params", id = "ModId",        editor = "text",  default = "" },
        { category = "Params", id = "OptionId",     editor = "text",  default = "" },
        { category = "Params", id = "OptionValues", editor = "table", default = {} },
    },
    visible = true,
}

function XModConfigEnum:Init()
    self.parent.idLabel:SetMargins(box(0, 0, self.MinWidth, 0))
    local current_value = ModConfig:Get(self.ModId, self.OptionId)
    local value_index = 1
    for i, option in ipairs(self.OptionValues) do
        if option.value == current_value then
            value_index = i
            break
        end
    end
    self:Set(value_index, false)
end

function XModConfigEnum:Set(page, update)
    if update == nil then update = true end
    self.current_page = page
    self.idPage:SetText(self.OptionValues[page].label)
    if update then
        ModConfig:Set(self.ModId, self.OptionId,
            self.OptionValues[page].value, ModConfig.internal_token)
    end
end

function XModConfigEnum:NextPage()
    local next_page = self.current_page + 1
    if next_page > #self.OptionValues then
        next_page = 1
    end
    self:Set(next_page)
end

function XModConfigEnum:PreviousPage()
    local next_page = self.current_page - 1
    if next_page < 1 then
        next_page = #self.OptionValues
    end
    self:Set(next_page)
end

----------------------------------------------------------------------------------------------------------
DefineClass.XModConfigNumberInput = {
    __parents = {
        "XWindow",
    },
    properties = {
        { category = "Params", id = "ModId",      editor = "text",   default = "" },
        { category = "Params", id = "OptionId",   editor = "text",   default = "" },
        { category = "Params", id = "OptionName", editor = "text",   default = "" },
        { category = "Params", id = "Min",        editor = "number", default = nil },
        { category = "Params", id = "Max",        editor = "number", default = nil },
        { category = "Params", id = "Step",       editor = "number", default = 1 },
    },
    HAlign = "right",
    LayoutMethod = "HList",
    TextFont = "OnScreenHintText",
    TextStyle = "OnScreenHintText",
}

function XModConfigNumberInput:Init()
    self.idRemove = XTextButton:new({
        Id = "idRemove",
        HAlign = "left",
        VAlign = "center",
        TextFont = "OnScreenHintText",
        TextStyle = "OnScreenHintText",
        MouseCursor = "UI/Cursors/Rollover.tga",
        FXMouseIn = "RocketRemoveCargoHover",
        RepeatStart = 300,
        RepeatInterval = 150,
        OnPress =  function(self) self.parent:Remove() end,
        Image = "UI/Infopanel/arrow_remove.tga",
        ColumnsUse = "abcc",
        RolloverTemplate = "Rollover",
    }, self)
    self.idAmount = XText:new({
        Id = "idAmount",
        Padding = box(2, 2, 5, 2),
        HAlign = "right",
        VAlign = "center",
        MinWidth = 60,
        MaxWidth = 70,
        TextFont = "UIPage",
        TextStyle = "UIPage",
        WordWrap = false,
        TextHAlign = "center",
        TextVAlign = "center",
    }, self)
    self.idAdd = XTextButton:new({
        Id = "idAdd",
        HAlign = "right",
        VAlign = "center",
        TextFont = "OnScreenHintText",
        TextStyle = "OnScreenHintText",
        MouseCursor = "UI/Cursors/Rollover.tga",
        FXMouseIn = "RocketRemoveCargoHover",
        RepeatStart = 300,
        RepeatInterval = 300,
        OnPress =  function(self) self.parent:Add() end,
        Image = "UI/Infopanel/arrow_add.tga",
        ColumnsUse = "abcc",
        RolloverTemplate = "Rollover",
    }, self)
    local AddRemoveRolloverHint = T{
        ModConfig.StringIdBase + 4,
        "<em><click></em> x<step><newline>"..
        "<em><shift> + <click></em> x<step10><newline>"..
        "<em><control> + <click></em> x<step100>",
        click = "<image "..MouseButtonImagesInText.MouseL..">",
        shift = ShortcutKeysToText({VKStrNames[const.vkShift]}),
        control = ShortcutKeysToText({VKStrNames[const.vkControl]}),
        step = self.Step,
        step10 = self.Step * 10,
        step100 = self.Step * 100,
    }
    self.idAmount:SetRolloverTextColor(RGB(255, 215, 0)) -- -- RolloverTextColor is Gold
    self.idRemove:SetRolloverTitle(self.OptionName)
    self.idAdd:SetRolloverTitle(self.OptionName)
    self.idRemove:SetRolloverText(self.OptionDesc or T{
        ModConfig.StringIdBase + 5,
        "<center>Decrease"
    })
    self.idAdd:SetRolloverText(self.OptionDesc or T{
        ModConfig.StringIdBase + 6,
        "<center>Increase"
    })
    self.idRemove:SetRolloverHint(AddRemoveRolloverHint)
    self.idAdd:SetRolloverHint(AddRemoveRolloverHint)
    self:Set(ModConfig:Get(self.ModId, self.OptionId))
    self.idAdd.MinWidth    = 20 -- Fix for Gagarin triangle issue
    self.idAdd.Columns     = 1  -- Fix for Gagarin triangle issue
    self.idAdd.ColumnsUse  = "aaaaa"  -- Fix for Gagarin triangle issue
    self.idRemove.MinWidth    = 20 -- Fix for Gagarin triangle issue
    self.idRemove.Columns     = 1  -- Fix for Gagarin triangle issue
    self.idRemove.ColumnsUse  = "aaaaa"  -- Fix for Gagarin triangle issue
end

function XModConfigNumberInput:Set(value, update)
    if update == nil then update = true end
    if type(value) ~= "number" then value = 0 end
    if self.Min ~= nil and value < self.Min then
        value = self.Min
    elseif self.Max ~= nil and value > self.Max then
        value = self.Max
    end
    self.current_value = value
    if value == self.Min then
        self.idRemove:SetVisible(false)
    else
        self.idRemove:SetVisible(true)
    end
    if value == self.Max then
        self.idAdd:SetVisible(false)
    else
        self.idAdd:SetVisible(true)
    end
    self.idAmount:SetText(LocaleInt(self.current_value))
    if update then
        ModConfig:Set(self.ModId, self.OptionId, value, ModConfig.internal_token)
    end
end

function XModConfigNumberInput:Remove()
    local step = self.Step
    if terminal.IsKeyPressed(const.vkShift) then
        step = step * 10
    elseif terminal.IsKeyPressed(const.vkControl) then
        step = step * 100
    end
    self:Set(self.current_value - step)
end

function XModConfigNumberInput:Add()
    local step = self.Step
    if terminal.IsKeyPressed(const.vkShift) then
        step = step * 10
    elseif terminal.IsKeyPressed(const.vkControl) then
        step = step * 100
    end
    self:Set(self.current_value + step)
end


----------------------------------------------------------------------------------------------------------
--  A bit of a shim class to emulate the behaviour of XScrollThumb from before the Curiosity update.
DefineClass.XModConfigScrollThumb = {
    __parents = {
        "XScrollThumb",
        "XEmbedIcon",
    },
}

function XModConfigScrollThumb:Init()
    -- The post-Curiosity XScrollThumb seems to be intended to be used only via its XSleekScroll
    -- subclass, which has an idThumb property that's functionally equivalent (for our purposes at
    -- least) to XEmbedIcon's idIcon property, so we just add an alias.
    self.idThumb = self.idIcon
end

DefineClass.XModConfigSlider = {
    __parents = {
        "XPropControl",
    },
    properties = {
        { category = "Params", id = "ModId",      editor = "text",   default = "" },
        { category = "Params", id = "OptionId",   editor = "text",   default = "" },
        { category = "Params", id = "OptionName", editor = "text",   default = "" },
        { category = "Params", id = "Min",        editor = "number", default = nil },
        { category = "Params", id = "Max",        editor = "number", default = nil },
        { category = "Params", id = "Step",       editor = "number", default = 1 },
    },
    RolloverOnFocus = true,
    FXMouseIn = "MenuItemHover",
}

function XModConfigSlider:Init()
    XWindow:new({
        Id = "idSliderTarget",
        OnScrollTo = function(self, scroll) self.parent:Set(scroll) end,
    }, self)
    self.slider = XModConfigScrollThumb:new({
        Dock = "right",
        VAlign = "center",
        Icon = "UI/Infopanel/progress_bar_slider.tga",
        Max = self.Max,
        Min = self.Min,
        StepSize = self.Step,
        Target = "idSliderTarget",
        MinWidth = 200,
        Horizontal = true,
    }, self)
    XFrame:new({
        ZOrder = 0,
        Image = "UI/Infopanel/progress_bar.tga",
        FrameBox = box(5, 0, 5, 0),
        SqueezeY = false,
    }, self.slider)
    self.idLabel = XText:new({
    	  TextFont = "OnScreenHintText",
        TextStyle = "OnScreenHintText",
        Translate = true,
        Margins = box(10, 0, 10, 0),
        Dock = "right",
    }, self)
    self:Set(ModConfig:Get(self.ModId, self.OptionId))
end

function XModConfigSlider:Set(value, update)
    if update == nil then update = true end
    if type(value) ~= "number" then value = 0 end
    if value < self.Min then
        value = self.Min
    elseif value > self.Max then
        value = self.Max
    end
    self.slider:SetScroll(value)
    if self.Label then
        self.idLabel:SetText(T{self.Label, value=value})
    end
    if update then
        ModConfig:Set(self.ModId, self.OptionId, value, ModConfig.internal_token)
    end
end


Code/MCR_3Options.lua
-- Code developed for Mod Config Reborn
-- Code originally developed by Waywocket as Mod Config
-- Modified to work with Gagarin Patch
-- Re-written to optimize code
-- Author @SkiRich
-- This mod is subject to the terms and conditions outlined in the LICENSE file included in this mod.
-- Created Oct 14th, 2018
-- Updated March 16th, 2021

-- StringIdBase this file 31-50

function OnMsg.ModConfigReady()
    local StringIdBase = ModConfig.StringIdBase

    -- Register this mod's name and description
    ModConfig:RegisterMod("_ModConfigReborn_", -- ID
        T{StringIdBase, "_Mod Config Reborn_"}, -- Optional display name, defaults to ID
        T{StringIdBase + 1, "Options for Mod Config Reborn"} -- Optional description
    )

    ModConfig:RegisterOption("_ModConfigReborn_", "ResetModConfigData", {
        name = T{StringIdBase + 2, "Reset All Mod Config Data:"},
        desc = T{StringIdBase + 3, "Reset and default all Mod Config Data. All mods will be defaulted if defaults defined. Do this only if you suspect registry corruption in mod config or for a fresh start."},
        type = "boolean",
        default = false
    })

end -- OnMsg.ModConfigReady()

local function MCRConfirmChangePopup(changetype)
	local changetypetext = {
		["blank"] = {text = "None Specified", explain = "will do nothing.", choice1 = "Close", choice2 = "Close"},
		["reset"] = {text = "Reset All Data", explain = "is not reversable.  All data will be defaulted for all mods.", choice1 = "Do It", choice2 = "Cancel"},
	}
	if not changetype then changetype = "blank" end
    CreateRealTimeThread(function()
        local params = {
            title = "Mod Config Reborn",
             text = string.format("The change <em>%s</em> %s", changetypetext[changetype].text, changetypetext[changetype].explain),
            choice1 = changetypetext[changetype].choice1,
            choice2 = changetypetext[changetype].choice2,
            image = "UI/Messages/death.tga",
            start_minimized = false,
        } -- params

        local choice = WaitPopupNotification(false, params)
        if changetype == "blank" then
        elseif changetype == "reset" then
        end


        if choice == 1 then
        	if changetype == "blank" then end
        	if changetype == "reset" then
        		ModConfig.CloseDialog()
        		ModConfig:ResetAllToDefaults() -- reset all data
        	end -- reset
        elseif choice == 2 then
        	if changetype == "blank" then end
        	if changetype == "reset" then end
        end -- if statement
    end ) -- CreateRealTimeThread
end -- MCRConfirmChangePopup()

function OnMsg.ModConfigChanged(mod_id, option_id, value, old_value, token)
    if g_ModConfigLoaded and mod_id == "_ModConfigReborn_" then
        -- reset mod config data
        if option_id == "ResetModConfigData" then
        	if value == true and token ~= "Reset" then
        		CreateRealTimeThread(function()
        			Sleep(100) -- Short delay to show the toggle
        			ModConfig:Toggle("_ModConfigReborn_", "ResetModConfigData", "Reset")
        			Sleep(100) -- Short Delay to allow process to happen
        			MCRConfirmChangePopup("reset") -- show confirm dialog
        		end) -- end RealTimeThread
        	end
        end
    end -- if g_ModConfigLoaded
end -- OnMsg.ModConfigChanged

metadata.lua
return PlaceObj('ModDef', {
	'title', "Mod Config Reborn",
	'description', "Mod Config Reborn\nv2.6.0  Sept 5th, 2021\nOriginal Code by Waywocket\nRebuilt by SkiRich\nContributions by ChoGGi",
	'image', "ModConfigRebornLogo.png",
	'last_changes', "v2.6.0  Sept 5th, 2021",
	'ignore_files', {
		"*.git/*",
		"*.svn/*",
		"*.gitignore",
		"*Images/*",
	},
	'id', "bpy1eJQ",
	'steam_id', "1542863522",
	'pops_desktop_uuid', "7ae484e3-d641-42ef-aac3-44041b721b1a",
	'pops_any_uuid', "24b1de3d-1f08-4ae5-b93f-de6bf11ca983",
	'author', "SkiRich",
	'version', 41,
	'lua_revision', 1007000,
	'saved_with_revision', 1007783,
	'code', {
		"Code/MCR_1Classes.lua",
		"Code/MCR_2Code.lua",
		"Code/MCR_3Options.lua",
	},
	'saved', 1631005874,
	'screenshot1', "Images/ModConfigRebornLogo.png",
	'screenshot2', "Images/ModConfigScreen.png",
	'screenshot3', "Images/ModConfigOption.png",
	'screenshot4', "Images/ModConfigOptions.png",
	'screenshot5', "Images/VoteUpApollo2.png",
	'TagInterface', true,
	'TagTools', true,
	'TagOther', true,
})